components/session-based-conditional-auth-functions/src/main/java/org/wso2/carbon/identity/extensions/authenticator/conditional/auth/functions/function/IsWithinSessionLimitFunction.java

```diff
@@ -0,0 +1,162 @@
+import org.wso2.carbon.identity.application.authentication.framework.exception.AuthenticationFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.exception.FrameworkException;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.identity.extensions.authenticator.conditional.auth.functions.util.SessionValidationConstants;
+import org.wso2.carbon.identity.extensions.authenticator.conditional.auth.functions.util.SessionValidationUtil;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+
+import static java.lang.Integer.parseInt;
+
+/**
+ * Represents javascript function provided in conditional authentication to check if the given user has valid number of
+ * sessions.
+ * The purpose is to perform dynamic authentication selection based on the active session count.
+ */
+public class IsWithinSessionLimitFunction implements IsValidFunction {
+
+    private static final Log log = LogFactory.getLog(IsWithinSessionLimitFunction.class);
+
+    /**
+     * Method to validate user session a given the authentication context and set of required attributes
+     *
+     * @param context Authentication context
+     * @param map     Hash map of attributes required for validation
+     * @return boolean value indicating the validation success/failure
+     * @throws AuthenticationFailedException when exception occurred in session retrieving method
+     */
+    @Override
+    public Boolean validate(JsAuthenticationContext context, Map<String, String> map)
+            throws AuthenticationFailedException {
+
+        boolean state = false;
+        int sessionLimit = getSessionLimitFromMap(map);
+        AuthenticatedUser authenticatedUser = context.getWrapped().getLastAuthenticatedUser();
+        if (authenticatedUser == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Unable to find the authenticated user from the Authentication context.");
+            }
+            throw new AuthenticationFailedException("Unable to find the Authenticated user from previous step");
+        }
+        try {
+            int sessionCount = getActiveSessionCount(authenticatedUser);
+            if (log.isDebugEnabled()) {
+                log.debug("Active session count: " + sessionCount + " and session limit : " + sessionLimit);
+            }
+            if (sessionCount < sessionLimit) {
+                state = true;
+            }
+        } catch (FrameworkException e) {
+            throw new AuthenticationFailedException("Problem occurred in session data retrieving", e);
+        } catch (NumberFormatException e) {
+            throw new AuthenticationFailedException("Failed to retrieve session count from response", e);
```

Comment:
Shall we make the error message more meaningful and add some context specific information, like authenticated user information ?

Reply:
username is added to both exception error messages